### PR TITLE
feat: made class to compute essential meta-data

### PR DESCRIPTION
**Pull request by tolgadur**
### Description 
We want to be able to filter the anonymized queries according to things such as physical cluster, organization and time created. For this we need to get/generate the relevant meta-data. The organization and physical cluster id is picked up from the control plane via a metric set [here](https://github.com/confluentinc/cc-spec-ksql/pull/233).

### Testing done 
Unit tests to make sure metadata gets computed as expected.

### References
jira: [KCI-342](https://confluentinc.atlassian.net/browse/KCI-342)
cc-structs-ksql update: [PR-223](https://github.com/confluentinc/cc-spec-ksql/pull/233)
cc-scheduler-service update: [PR-2240](https://github.com/confluentinc/cc-scheduler-service/pull/2240)
protobuff update: [PR-1507](https://github.com/confluentinc/cc-structs/pull/1507)
old pr that was closed accidentally: [PR-7425](https://github.com/confluentinc/ksql/pull/7425)

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksqldb-parser/src/main/java/io/confluent/ksql/util/QueryMetaDataCollector.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+package io.confluent.ksql.util;
+
+import com.google.common.collect.ImmutableMap;
+import java.nio.charset.StandardCharsets;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.UUID;
+import org.joda.time.DateTime;
+
+public final class QueryMetaDataCollector {
```
Comment (swist) [inline ksqldb-parser/src/main/java/io/confluent/ksql/util/QueryMetaDataCollector.java:25]: This isn't the `Collector` pattern - the name is misleading: https://www.codeproject.com/Articles/1076386/Collector-Translator-And-Formatter-Pattern

Sounds like this class is a `QueryMetadataBuilder`, not collector

File: ksqldb-parser/src/main/java/io/confluent/ksql/util/QueryMetaDataCollector.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+  private QueryMetaDataCollector() {}
+
+  public static ImmutableMap<String, String> buildMetaData(
```
Comment (swist) [inline ksqldb-parser/src/main/java/io/confluent/ksql/util/QueryMetaDataCollector.java:29]: This should return a QueryMetadata and method should be called:

```QueryMetadata.of(```

File: ksqldb-parser/src/main/java/io/confluent/ksql/util/QueryMetaDataCollector.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    final Map<String, String> metaData = new HashMap<>();
+    final String clusterNamespace = config
+        .getString(KsqlConfig.KSQL_CCLOUD_QUERYANONYMIZER_CLUSTER_NAMESPACE);
+
+    metaData.put("cluster_namespace", clusterNamespace);
+    metaData.put("id", getQueryId(nonAnonQuery, clusterNamespace));
+    metaData.put("structurally_similar_id", getStructurallySimilarId(anonQuery, clusterNamespace));
+    metaData.put("time_of_creation", DateTime.now().toString());
+
+    return ImmutableMap.copyOf(metaData);
```
Comment (swist) [inline ksqldb-parser/src/main/java/io/confluent/ksql/util/QueryMetaDataCollector.java:44]: use the `ImmutableMap.builder` instead

File: ksqldb-parser/src/main/java/io/confluent/ksql/util/QueryMetaDataCollector.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+import java.util.Map;
+import java.util.UUID;
+import org.joda.time.DateTime;
+
+public final class QueryMetaDataCollector {
+
+  private QueryMetaDataCollector() {}
+
+  public static ImmutableMap<String, String> buildMetaData(
+      final KsqlConfig config,
```
Comment (swist) [inline ksqldb-parser/src/main/java/io/confluent/ksql/util/QueryMetaDataCollector.java:30]: First arg should be  `String namespace` rather than KsqlConfig so that you don't tight couple this with configuration stack

File: ksqldb-parser/src/test/java/io/confluent/ksql/util/QueryMetaDataCollectorTest.java [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+import org.junit.Test;
+
+public class QueryMetaDataCollectorTest {
+  private KsqlConfig ksqlConfig;
+
+  @Before
+  public void setUp() {
+    // initialize ksql config
+    Map<String, String> ksqlConfigMap = new HashMap<>();
+    ksqlConfigMap.put(KsqlConfig.KSQL_CCLOUD_QUERYANONYMIZER_CLUSTER_NAMESPACE,
```
Comment (swist) [inline ksqldb-parser/src/test/java/io/confluent/ksql/util/QueryMetaDataCollectorTest.java:31]: see above re first arg: then you don't have to do this complicated mocking, just pass the literal around and your testing becomes easier

File: ksqldb-parser/src/main/java/io/confluent/ksql/util/QueryMetaDataCollector.java [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+  }
+
+  private static String getQueryId(final String query, final String namespace) {
+    final String genericQuery = getGenericQueryForm(query);
+    final String namespacePlusQuery = namespace + genericQuery;
+
+    return UUID.nameUUIDFromBytes(namespacePlusQuery.getBytes(StandardCharsets.UTF_8)).toString();
+  }
+
+  private static String getStructurallySimilarId(final String query, final String namespace) {
```
Comment (swist) [inline ksqldb-parser/src/main/java/io/confluent/ksql/util/QueryMetaDataCollector.java:54]: You don't need this method, It's the same as calling: 

```getQueryId(anonQuery, clusterNamespace)```

File: ksqldb-parser/src/test/java/io/confluent/ksql/util/QueryMetaDataCollectorTest.java [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+        "testData.testData");
+    ksqlConfig = new KsqlConfig(ksqlConfigMap);
+  }
+
+  @Test
+  public void shouldGetCorrectClusterNamespaceData() {
```
Comment (swist) [inline ksqldb-parser/src/test/java/io/confluent/ksql/util/QueryMetaDataCollectorTest.java:37]: You need tests for edge case behaviour:
1. empty query
2. empty namespace
3. empty anon query

**Review by swist — CHANGES_REQUESTED** [review]

File: ksqldb-parser/src/main/java/io/confluent/ksql/util/QueryMetaDataCollector.java (hunk 2 above)
Comment (swist) [reply ksqldb-parser/src/main/java/io/confluent/ksql/util/QueryMetaDataCollector.java:29]: Or you could make this a QueryMetadata class with an `of` constructor

File: ksqldb-parser/src/main/java/io/confluent/ksql/util/QueryMetaDataCollector.java [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+      final String anonQuery) {
+
+    // get metadata from control plane
+    final Map<String, String> metaData = new HashMap<>();
+    final String clusterNamespace = config
+        .getString(KsqlConfig.KSQL_CCLOUD_QUERYANONYMIZER_CLUSTER_NAMESPACE);
+
+    metaData.put("cluster_namespace", clusterNamespace);
+    metaData.put("id", getQueryId(nonAnonQuery, clusterNamespace));
+    metaData.put("structurally_similar_id", getStructurallySimilarId(anonQuery, clusterNamespace));
```
Comment (swist) [inline ksqldb-parser/src/main/java/io/confluent/ksql/util/QueryMetaDataCollector.java:41]: Something like `anonymisedQueryId` would be a good name here I reckon

File: ksqldb-parser/src/main/java/io/confluent/ksql/util/QueryGuid.java [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.util;
+
+import java.nio.charset.StandardCharsets;
+import java.util.UUID;
+import org.joda.time.DateTime;
```
Comment (swist) [inline ksqldb-parser/src/main/java/io/confluent/ksql/util/QueryGuid.java:20]: Is that the DateTime library that we use throughout?

**Review by swist — APPROVED** [review]
lgtm, just one question

File: ksqldb-parser/src/main/java/io/confluent/ksql/util/QueryGuid.java (hunk 9 above)
Comment (tolgadur) [reply ksqldb-parser/src/main/java/io/confluent/ksql/util/QueryGuid.java]: You're right. I important the more frequently used one.
